### PR TITLE
CDPCAM-48, CDPCAM-43: Custom output and links to docs.

### PR DESCRIPTION
**Pull request by wonderslug**
This commits provides the capability to do different output types
(tex and json for now) and provides the different issues encountered
with the ability to have links and different formatting.

It does this through centralizing the Issue messages into
issue_templates.yml files which get loaded at runtime to present the issues
to the renderers.  The renderers pick the format mechanism defined by the
render_type in the issue_templates.yml.

The issue ids can be generated into constants using the
scripts/update_issue_templates.py script.  This will also be run during
the pre-commit.

Signed-off-by: Brian Towles <btowles@cloudera.com>

### CODE REVIEW

File: cdpctl/command/validate.py [hunk 1]
```diff
@@ -50,11 +50,18 @@
...
+from cdpctl.validation.renderer import get_renderer
 
 
-def run_validation(target: str, config_file: str, debug: bool = False) -> None:
+def run_validation(
+    target: str,
+    config_file: str,
+    debug: bool = False,
+    output_format: str = "text",
+    output_file: str = "-",
```
Comment (aakanser) [inline cdpctl/command/validate.py:63]: why not blank as default?

File: cdpctl/validation/aws_utils.py [hunk 2]
```diff
@@ -63,32 +71,25 @@ def get_client(client_type: str, config):
...
-        ),
+        key_not_found_issue=AWS_REGION_CONFIG_NOT_DEFINED,
+        data_expected_issue=AWS_REGION_NOT_DEFINED,
     )
 
-    if profile_name:
-        session = boto3.session.Session(profile_name=profile_name)
-        return session.client(client_type)
     if region_name:
+        if profile_name:
```
Comment (aakanser) [inline cdpctl/validation/aws_utils.py:86]: could this be combined with line above?

File: cdpctl/validation/__init__.py [hunk 3]
```diff
@@ -57,14 +70,210 @@ class UnrecoverableValidationError(ValidationError):
...
+        }
+    get_issues()[context.validation_name][issue_type.value].append(issue)
+
+
+def fail(
+    template: str, subjects: List[str] = None, resources: List[str] = None
+) -> None:
+    """Fail the validation."""
+    if template not in _issue_templates:
+        raise UnrecoverableValidationError(f"Unable to find issue template {template}")
```
Comment (aakanser) [inline cdpctl/validation/__init__.py:238]: This would make sense to us, but for a customer running this tool, is there a better way to fail here? Will other validations run if we throw unrecoverable error here?

File: cdpctl/validation/aws_utils.py [hunk 4]
```diff
@@ -198,11 +195,11 @@ def get_role(
...
-        pytest.fail(role_missing_message.format(role_name), False)
+        fail(template=missing_issue, resources=[role_name])
     except iam_client.exceptions.ServiceFailureException as e:
-        raise Exception(service_failure_message) from e
+        raise UnrecoverableValidationError(service_failure_message) from e
     # handling stub client error
-    except ClientError:
-        pytest.fail(role_missing_message.format(role_name), False)
+    except ClientError as ce:
+        fail(missing_issue, ce.args[0])
```
Comment (aakanser) [inline cdpctl/validation/aws_utils.py:203]: why not use `role_name` here instead of ce.args? What does that look like?

Comment (aakanser) [reply cdpctl/validation/aws_utils.py:203]: Is this always a comm error? Hence no role in error?

File: cdpctl/validation/conftest.py [hunk 5]
```diff
@@ -128,17 +133,13 @@ def pytest_sessionfinish(
...
-        if session.issues:
-            click.secho("\n--- Issues Found ---", fg="yellow")
-            # pylint: disable=unused-variable
-            for item, result in session.issues.items():
-                click.echo(f"- {result}")
-                click.echo("\n")
-            click.secho("-------------------", fg="yellow")
-        else:
-            click.secho("No Issues Found", fg="green")
+        # env = Environment(
```
Comment (aakanser) [inline cdpctl/validation/conftest.py:137]: delete?

File: cdpctl/validation/infra/issue_templates.yml [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+summary: "The s3a url {0} is invalid."
+---
+id: AWS_S3_BUCKET_DOES_NOT_EXIST
+summary: "S3 bucket {0} does not exist."
+---
+id: AWS_S3_BUCKET_FORBIDDEN_ACCESS
+summary: "S3 bucket {0} has forbidden access."
+---
+id: AWS_NON_CCM_DEFAULT_SG_NEEDS_TO_ALLOW_CDP_CIDRS
+summary: "When not using CCM (tunnel = false), the default security group {0} is missing TCP port 9443 to be allowed access from the following Cloudera CDP CIDRs"
```
Comment (aakanser) [inline cdpctl/validation/infra/issue_templates.yml:72]: maybe rephrase to: 
The default security group {0} is missing TCP port 9443 for the following Cloudera CDP CIDRs when not using CCM (env:tunnel = false).

File: cdpctl/validation/infra/issue_templates.yml [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+render_type: list
+---
+id: AWS_DEFAULT_SG_NEEDS_ALLOW_ACCESS_INTERNAL_TO_VPC
+summary: "Your Default Security Group {0} should allow access to all TCP and UDP ports (0-65535) internal to the VPC"
+---
+id: AWS_GATEWAY_SG_NEEDS_ALLOW_ACCESS_INTERNAL_TO_VPC
+summary: "Your Gateway Security Group {0} should allow access to all TCP and UDP ports (0-65535) internal to the VPC"
+---
+id: AWS_SSH_KEY_ID_DOES_NOT_EXIST
+summary: "SSH key id ({0}) do not exist."
```
Comment (aakanser) [inline cdpctl/validation/infra/issue_templates.yml:86]: *does

File: cdpctl/validation/infra/issue_templates.yml [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+---
+id: AWS_SSH_IS_INVALID
+summary: "SSH Key ID ({0}) is not valid."
+---
+id: AWS_NOT_ENOUGH_SUBNETS_PROVIDED
+summary: "Not enough {0} Subnets provided, at least 3 subnets required."
```
Comment (aakanser) [inline cdpctl/validation/infra/issue_templates.yml:92]: I couldn't find where this is used. Is {0} going to be the number of subnets found?

File: cdpctl/validation/__init__.py (hunk 3 above)
Comment (wonderslug) [reply cdpctl/validation/__init__.py:238]: No this is an immediate error only for us.  The customer should not see this.  UnrecoverableValidationError means it stops.

File: cdpctl/validation/infra/issue_templates.yml (hunk 8 above)
Comment (aakanser) [reply cdpctl/validation/infra/issue_templates.yml:92]: If it's just a placeholder for now, that's fine.

File: cdpctl/command/validate.py (hunk 1 above)
Comment (wonderslug) [reply cdpctl/command/validate.py:63]: because convention for Unix usually is - meaning stdout.  for tar etc.

File: cdpctl/validation/infra/issue_templates.yml (hunk 8 above)
Comment (wonderslug) [reply cdpctl/validation/infra/issue_templates.yml:92]: No for that message its allowing us to pass in Public or Private to reuse

Comment (wonderslug) [reply cdpctl/validation/infra/issue_templates.yml:92]: Used in validate_aws_subnets.py line 91

File: cdpctl/validation/infra/issue_templates.yml (hunk 7 above)
Comment (wonderslug) [reply cdpctl/validation/infra/issue_templates.yml:86]: done

File: cdpctl/validation/renderer/__init__.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        """Render the issues found as a json format."""
+
+        json_issues = []
+        for key, value in issues.items():
+            json_rep = {}
+            json_rep["validation"] = key
+            json_rep["problems"] = []
+            json_rep["warnings"] = []
+            for problem in value["problem"]:
+                json_prob = {}
```
Comment (aakanser) [inline cdpctl/validation/renderer/__init__.py:81]: this could be:
  json_rep["problems"].append({"message": ..., "resources": ...})

File: cdpctl/validation/infra/issue_templates.yml (hunk 6 above)
Comment (wonderslug) [reply cdpctl/validation/infra/issue_templates.yml:72]: done

File: cdpctl/validation/renderer/__init__.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+                json_prob["message"] = problem.message
+                json_prob["resources"] = problem.resources
+                json_rep["problems"].append(json_prob)
+            for warning in value["warning"]:
+                json_warn = {}
```
Comment (aakanser) [inline cdpctl/validation/renderer/__init__.py:86]: same as above

File: cdpctl/validation/conftest.py (hunk 5 above)
Comment (wonderslug) [reply cdpctl/validation/conftest.py:137]: done

File: cdpctl/validation/aws_utils.py (hunk 4 above)
Comment (wonderslug) [reply cdpctl/validation/aws_utils.py:203]: yeah mistype

File: scripts/templates/issues.py.j2 [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+# AND WILL NOT DEFEND, INDEMNIFY, NOR HOLD YOU HARMLESS FOR ANY CLAIMS
+# ARISING FROM OR RELATED TO THE CODE; AND (D)WITH RESPECT TO YOUR EXERCISE
+# OF ANY RIGHTS GRANTED TO YOU FOR THE CODE, CLOUDERA IS NOT LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, PUNITIVE OR
+# CONSEQUENTIAL DAMAGES INCLUDING, BUT NOT LIMITED TO, DAMAGES
+# RELATED TO LOST REVENUE, LOST PROFITS, LOSS OF INCOME, LOSS OF
+# BUSINESS ADVANTAGE OR UNAVAILABILITY, OR LOSS OR CORRUPTION OF
+# DATA.
+#
+# Source File Name: issues.py
```
Comment (aakanser) [inline scripts/templates/issues.py.j2:40]: issues.py.j2?

File: tests/validation/infra/test_validate_aws_subnets.py [hunk 12]
```diff
@@ -211,45 +209,6 @@ def test_aws_public_subnets_validation_failure(
         func(config, ec2_client)
 
 
-def test_aws_public_subnets_suffix_validation_success(
```
Comment (aakanser) [inline tests/validation/infra/test_validate_aws_subnets.py:214]: were these validations removed?

File: cdpctl/validation/infra/issue_templates.yml (hunk 8 above)
Comment (aakanser) [reply cdpctl/validation/infra/issue_templates.yml:92]: Nice!

**Review by aakanser — APPROVED** [review]
looks good, left a few comments

File: scripts/templates/issues.py.j2 (hunk 11 above)
Comment (wonderslug) [reply scripts/templates/issues.py.j2:40]: This is the template to output the issues.py from the scripts/update_issue_templates.py

File: tests/validation/infra/test_validate_aws_subnets.py (hunk 12 above)
Comment (wonderslug) [reply tests/validation/infra/test_validate_aws_subnets.py:214]: Because we are not doing suffix validations.  The validations were skipped and not needed.  If we need to add them back later we can.

File: cdpctl/validation/infra/__init__.py [hunk 13]
```diff
@@ -39,5 +38,5 @@
 #
 # Source File Name:  __init__.py
 ###
-"""This is empty to fix package for VSCode."""
+"""Base infa package."""
```
Comment (mahendra-rk) [inline cdpctl/validation/infra/__init__.py:41]: nit: Typo

File: cdpctl/validation/aws_utils.py (hunk 2 above)
Comment (wonderslug) [reply cdpctl/validation/aws_utils.py:86]: No we still want the return for the region without a profile name.  The region has to be defined.  Region and profile CAN be defined and returns a different client.

File: cdpctl/validation/infra/issue_templates.yml [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+id: AWS_SUBNETS_DO_NOT_EXIST
+summary: "The following {0} Subnets do not exist."
+render_type: list
+---
+id: AWS_NOT_ENOUGH_AZ_FOR_SUBNETS
+summary: "Not enough availability zones, {0} subnets should spread across at least 2 availability zones."
+---
+id: AWS_SUBNETS_WITHOUT_INTERNET_GATEWAY
+summary: "These {0} Subnets do not have an internet gateway(s)"
+docs_link: https://docs.cloudera.com/cdp/latest/requirements-aws/topics/mc-aws-req-vpc.html
```
Comment (mahendra-rk) [inline cdpctl/validation/infra/issue_templates.yml:112]: adding link markup's `[link=URL]text[/link]`. thoughts?

File: cdpctl/validation/infra/issues.py [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+AWS_INVALID_DATA = "AWS_INVALID_DATA"
+AWS_SUBNETS_DO_NOT_EXIST = "AWS_SUBNETS_DO_NOT_EXIST"
+AWS_NOT_ENOUGH_AZ_FOR_SUBNETS = "AWS_NOT_ENOUGH_AZ_FOR_SUBNETS"
+AWS_SUBNETS_WITHOUT_INTERNET_GATEWAY = "AWS_SUBNETS_WITHOUT_INTERNET_GATEWAY"
+AWS_SUBNETS_OR_VPC_WITHOUT_INTERNET_GATEWAY = "AWS_SUBNETS_OR_VPC_WITHOUT_INTERNET_GATEWAY"
+AWS_SUBNETS_WITHOUT_VALID_RANGE = "AWS_SUBNETS_WITHOUT_VALID_RANGE"
+AWS_SUBNETS_MISSING_K8S_LB_TAG = "AWS_SUBNETS_MISSING_K8S_LB_TAG"
+AWS_SUBNETS_NOT_PART_OF_VPC = "AWS_SUBNETS_NOT_PART_OF_VPC"
+AWS_DNS_SUPPORT_NOT_ENABLED_FOR_VPC = "AWS_DNS_SUPPORT_NOT_ENABLED_FOR_VPC"
+AWS_VPC_NOT_FOUND_IN_ACCOUNT = "AWS_VPC_NOT_FOUND_IN_ACCOUNT"
```
Comment (mahendra-rk) [inline cdpctl/validation/infra/issues.py]: nit: missing blank line at the end

File: cdpctl/validation/renderer/__init__.py (hunk 9 above)
Comment (wonderslug) [reply cdpctl/validation/renderer/__init__.py]: done

File: cdpctl/validation/renderer/__init__.py (hunk 10 above)
Comment (wonderslug) [reply cdpctl/validation/renderer/__init__.py]: done

File: cdpctl/validation/renderer/__init__.py [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+class JsonValidationRenderer(ValidationRenderer):
+    def render(self, issues, output_file):
+        """Render the issues found as a json format."""
+
+        json_issues = []
+        for key, value in issues.items():
+            json_rep = {}
+            json_rep["validation"] = key
+            json_rep["problems"] = []
+            json_rep["warnings"] = []
```
Comment (mahendra-rk) [inline cdpctl/validation/renderer/__init__.py:85]: single line would be better ? `json_rep = {"validation": key, "problems" : [] , "warnings": []}`

File: cdpctl/validation/infra/issue_templates.yml (hunk 14 above)
Comment (wonderslug) [reply cdpctl/validation/infra/issue_templates.yml:112]: Won't work for text output.  We should leave the markups to the renderers.  one may be HTML which would do it differently.

File: cdpctl/validation/renderer/__init__.py (hunk 16 above)
Comment (wonderslug) [reply cdpctl/validation/renderer/__init__.py:85]: already commented on and fixed

File: scripts/templates/issues.py.j2 (hunk 11 above)
Comment (mahendra-rk) [inline scripts/templates/issues.py.j2:40]: file name `issues.py.j2`

**Review by mahendra-rk — APPROVED** [review]
LGTM!

File: scripts/templates/issues.py.j2 (hunk 11 above)
Comment (wonderslug) [reply scripts/templates/issues.py.j2:40]: This is the template for the jinja output from update_issue_templates.py. hence j2
